### PR TITLE
KEP-330: Checkpoints in pbft, refactoring in pbft tests

### PR DESCRIPTION
**Pull request by isabelsavannah**

### CODE REVIEW

File: pbft/pbft.cpp [hunk 1]
```diff
@@ -469,4 +491,130 @@ pbft::handle_failure()
...
+
+    this->stable_checkpoint = cp;
+    this->stable_checkpoint_proof = this->unstable_checkpoint_proofs[cp];
+
+    LOG(info) << boost::format("Checkpoint %1% at seq %2% is now stable; clearing old data")
+              % cp.second
+              % cp.first;
+
+    // Remove obsolete local checkpoints
+    auto local_start = this->local_unstable_checkpoints.begin();
```
Comment (ebruck) [inline pbft/pbft.cpp]: I tend to const these type of variables.

File: pbft/pbft.cpp [hunk 2]
```diff
@@ -469,4 +491,130 @@ pbft::handle_failure()
...
+    // Iterator to the first unstable checkpoint that's newer than this one. This logic assumes that CHECKPOINT_INTERVAL
+    // is >= 2, otherwise we would have do do something awkward here
+    auto local_end = this->local_unstable_checkpoints.upper_bound(checkpoint_t(cp.first+1, ""));
+    size_t local_removed = std::distance(local_start, local_end);
```
Comment (ebruck) [inline pbft/pbft.cpp]: same

File: pbft/pbft.cpp [hunk 3]
```diff
@@ -469,4 +491,130 @@ pbft::handle_failure()
...
+
+
+    // Remove obsolete checkpoint proof collections
+    auto start = this->unstable_checkpoint_proofs.begin();
+    auto end = this->unstable_checkpoint_proofs.upper_bound(checkpoint_t(cp.first+1, ""));
+    size_t to_remove = std::distance(start, end);
+    this->unstable_checkpoint_proofs.erase(start, end);
+    LOG(debug) << boost::format("Cleared %1% unstable checkpoint proof sets") % to_remove;
+
+
```
Comment (ebruck) [inline pbft/pbft.cpp]: There's a few things going on in here. Is it worth breaking out to separate functions?

File: pbft/pbft_service_base.hpp [hunk 4]
```diff
@@ -74,13 +74,24 @@ namespace bzn
...
          * A checkpoint has been stabilized, so we no longer need any history from before then.
          */
         virtual void consolidate_log(uint64_t sequence_number) = 0;
 
-        virtual ~pbft_service_base() = default;
+        /*
+         * Callback when a request is executed (not committed, since the service is responsible for the difference
+         * between the two). Should only be called once for each sequence number, in strictly increasing order.
+         */
+        virtual void register_execute_handler(std::function<void(const pbft_request&, uint64_t)> handler) = 0;
```
Comment (ebruck) [inline pbft/pbft_service_base.hpp]: create "using" for handler type

File: pbft/pbft.cpp [hunk 5]
```diff
@@ -83,6 +85,23 @@ pbft::start()
                 this->audit_heartbeat_timer->async_wait(
                         std::bind(&pbft::handle_audit_heartbeat_timeout, shared_from_this(), std::placeholders::_1));
 
+                this->service->register_execute_handler(
+                        [weak_this = this->weak_from_this(), fd = this->failure_detector]
+                                (const pbft_request& req, uint64_t sequence)
+                                        {
+                                            fd->request_executed(req);
+
+                                            if (sequence % CHECKPOINT_INTERVAL == 0)
```
Comment (ebruck) [inline pbft/pbft.cpp]: Is there an error if weak_this is not around anymore? 
Should there be logging at least?

File: pbft/pbft.cpp [hunk 6]
```diff
@@ -469,4 +491,130 @@ pbft::handle_failure()
...
+    // Iterator to the first unstable checkpoint that's newer than this one. This logic assumes that CHECKPOINT_INTERVAL
+    // is >= 2, otherwise we would have do do something awkward here
+    auto local_end = this->local_unstable_checkpoints.upper_bound(checkpoint_t(cp.first+1, ""));
+    size_t local_removed = std::distance(local_start, local_end);
+    this->local_unstable_checkpoints.erase(local_start, local_end);
+    LOG(debug) << boost::format("Cleared %1% unstable local checkpoints") % local_removed;
+
+
+    // Remove obsolete checkpoint proof collections
+    auto start = this->unstable_checkpoint_proofs.begin();
```
Comment (ebruck) [inline pbft/pbft.cpp]: const

File: include/boost_asio_beast.hpp [hunk 7]
```diff
@@ -113,8 +113,11 @@ namespace bzn::asio
 
         virtual std::unique_ptr<bzn::asio::strand_base> make_unique_strand() = 0;
 
+        virtual void post(std::function<void()>) = 0;
```
[the remark this answers is not in the document]
Comment (isabelsavannah) [reply include/boost_asio_beast.hpp]: Do you mean an alias for the parameter type? If not then I don't understand what you're asking.

**Pull request merged**